### PR TITLE
CCTBX

### PR DESCRIPTION
**Pull request by bkpoon**
This adds the Computational Crystallography Toolbox as a conda package.

Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by Anthchirp** [comment]
license file packaged here is different from [license file in the original package](https://github.com/cctbx/cctbx_project/blob/master/LICENSE.txt). Latter is not BSD-3-clause.

**Comment by bkpoon** [comment]
Waiting for merging of some extra code into master, a cctbx release (https://github.com/cctbx/cctbx_project/issues/451), and some user feedback on some test packages in the `cctbx-dev` channel (http://phenix-online.org/pipermail/cctbxbb/2020-March/002202.html, http://phenix-online.org/pipermail/phenixbb/2020-March/024564.html).

**Comment by bkpoon** [comment]
Finalizing some installation scripts and the official release version

**Comment by Anthchirp** [comment]
Wow. Well done. Looking at the commit history this was a real tour de force. I'm looking forward to having a detailled look at the weekend, but as a quick immediate comment: the branch base commit seems to be about 9 months old. Could you merge the current `master` into your branch just to make sure that the build scripts are up to date?

**Comment by bkpoon** [comment]
@Anthchirp I brought the branch up to date, but it should not matter since the [pull request trigger](https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#pr-triggers) for Azure Pipelines should run on the merged code.

The packages from the latest build are downloadable artifacts, and I have uploaded them to the `cctbx-dev` channel for some lower level testing. You can install them into your active environment with
```
conda install -c cctbx-dev cctbx
```
Like `matplotlib`, there is a `cctbx-base` package that does not have the `wxpython` and `pyside2` dependencies. This is useful for users who want to use `jupyterlab` or do not use any of the GUI features.

**Comment by bkpoon** [comment]
@conda-forge/staged-recipes This took a while to get everything sorted out, but this recipe should be ready for review.

Some notes:

1. The `yum_requirements.txt` is for some OpenGL headers and libraries. We have some GUI code that builds and links against these dependencies. I use the `missing_dso_whitelist` to ignore the libGL and libGLU shared libraries. Otherwise, I get an error on linux.
2. I am not sure if I am pinning `numpy` correctly. Should the version be 1.16 instead of 1.19? The `boost` pinning seems to be fine (1.72 is used instead of 1.74)?
3. Is it possible to enable `linux_ppc64le`? It's not really necessary right now, but some users run this code on the [Summit HPC system](https://www.olcf.ornl.gov/olcf-resources/compute-systems/summit/) which uses Power9 processors. Currently, we are using the ppc `conda` packages for dependencies and then compiling the code.

Thanks!

**Comment by chrisburr** [comment]
For 3: After this is merged you should open a PR to add it to: https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/migrations/arch_rebuild.txt

The bots will then take care of migrating it.

File: recipes/cctbx/LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,44 @@
+*** License agreement ***
```
Comment (chrisburr) [inline recipes/cctbx/LICENSE.txt:1]: No need for this file, conda build will find it from the source tarball.

File: recipes/cctbx/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  missing_dso_whitelist:
+    - '*/libGL.so.1'  # [linux]
+    - '*/libGLU.so.1'  # [linux]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+  host:
+    - boost >=1.65
```
Comment (chrisburr) [inline recipes/cctbx/meta.yaml:33]: You shouldn't set a version constraint here as it will come from the global pinning. As it stands it will always build against the newest boost which will make conflicts very likely.

Unrelated, are you using boost python? If not you should use boost-cpp here.

**Comment by Anthchirp** [comment]
> @Anthchirp I brought the branch up to date, but it should not matter since the [pull request trigger](https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#pr-triggers) for Azure Pipelines should run on the merged code.

I can't speak to that. I just picked that up [from the official recommendation](https://github.com/conda-forge/staged-recipes/pull/11288#issuecomment-692211722) 🙂 

> The yum_requirements.txt is for some OpenGL headers and libraries. We have some GUI code that builds and links against these dependencies. I use the missing_dso_whitelist to ignore the libGL and libGLU shared libraries. Otherwise, I get an error on linux.

Have you seen the [libGL section in the knowledgebase](https://conda-forge.org/docs/maintainer/knowledge_base.html#libgl) advising to use CDT packages?

> I am not sure if I am pinning numpy correctly. Should the version be 1.16 instead of 1.19? The boost pinning seems to be fine (1.72 is used instead of 1.74)?

As I understand it (but I don't do numpy packages), if you just pick numpy directly you'll build against the newest version, meaning that it will only work with the newest version. So you'll need to compile against the oldest supported version. Again, [the knowledgebase might help out here.](https://conda-forge.org/docs/maintainer/knowledge_base.html#building-against-numpy)

File: recipes/cctbx/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  sha256: 38b41bb5e3b968e499a35c30de81147abb946c8de2728e90676f24843348b033
+  patches:
+    - crys3d.patch
+    - cbf.patch  # [win]
+    - libtbx.patch
+
+build:
+  number: 0
+  missing_dso_whitelist:
+    - '*/libGL.so.1'  # [linux]
```
Comment (chrisburr) [inline recipes/cctbx/meta.yaml:18]: See: https://conda-forge.org/docs/maintainer/knowledge_base.html#libgl

File: recipes/cctbx/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    - {{ compiler('cxx') }}
+  host:
+    - boost >=1.65
+    - boost-cpp
+    - eigen
+    - future
+    - hdf5
+    - libtiff
+    - msgpack-c
+    - numpy >=1.14
```
Comment (chrisburr) [inline recipes/cctbx/meta.yaml]: This should be empty so the version from conda forge pinning is used instead of the latest.

In staged recipes you might find it doesn't work properly, if so set the versions to match conda-forge-pinning and then delete them once the feedstock is made (it's important to rerender after removing them).

**Comment by bkpoon** [comment]
@conda-forge-admin please rerender

**Review by chrisburr — COMMENTED** [review]
Apologies, I have to dash now so there might be more but this is a start at reviewing.

**Comment by bkpoon** [comment]
@Anthchirp Thanks! I forgot where that link was. Let's see what happens now.

File: recipes/cctbx/meta.yaml (hunk 2 above)
Comment (bkpoon) [reply recipes/cctbx/meta.yaml]: Yes, we are using Boost.Python to expose C++ code to Python. And just committed a change.

File: recipes/cctbx/LICENSE.txt (hunk 1 above)
Comment (bkpoon) [reply recipes/cctbx/LICENSE.txt]: Just committed a change.

File: recipes/cctbx/meta.yaml (hunk 3 above)
Comment (bkpoon) [reply recipes/cctbx/meta.yaml:18]: Just committed a change.

File: recipes/cctbx/meta.yaml (hunk 4 above)
Comment (bkpoon) [reply recipes/cctbx/meta.yaml]: Just committed a change.

**Comment by bkpoon** [comment]
@conda-forge-admin please rerender

**Comment by bkpoon** [comment]
@chrisburr Thanks! I've made the changes and rerendered.

File: recipes/cctbx/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        - wxpython
+    test:
+      imports:
+        - scitbx.array_family.flex
+
+about:
+  home: https://github.com/cctbx/cctbx_project
+  license: BSD-3-Clause
+  license_family: BSD
+  license_file: LICENSE.txt
```
Comment (Anthchirp) [inline recipes/cctbx/meta.yaml:85]: This is actually not true. Neither cctbx nor the enclosed license file are BSD-3-Clause. Granted, this is a can of worms, but worth getting it right. (Usual disclaimer: I'm not a lawyer.)

The license file contains an additional derivate works provision, making it not BSD-3.

Further, this package is vendoring other source code with different licenses that are not acknowledged. A cursory sweep found:

* boost licenses in 
  * `site-packages/boost_adaptbx/graph/breadth_first_search.py`
  * `site-packages/scitbx/auto_array.h`
  * `site-packages/scitbx/random/mersenne_twister.h`
* LGPL
  * `site-packages/ccp4io/`
  * `site-packages/cctbx/uctbx/determine_unit_cell/lgpl.txt`
  * `site-packages/libtbx/clusterTests.py`
  * `site-packages/libtbx/cluster.py`
* MIT
  * `site-packages/libtbx/objgraph.py`
* STL in
  * `site-packages/scitbx/array_family/shared_plain.h`
* custom licenses in 
  * `site-packages/chiltbx`
  * `site-packages/libtbx/fastentrypoints.py`
  * `site-packages/libtbx/xmlrpc_utils.py`
  * `site-packages/scitbx/math/parabolic_cylinder_d.h`
  * `site-packages/ucif/cifParser.h` 
  * `site-packages/wxtbx/checklistctrl.py`
  * `site-packages/wxtbx/metallicbutton.py`
  * `site-packages/wxtbx/sequence_view.py`
* "Copyright (C) Diamond Light Source", some of which are declared BSD, some are not.
  * `site-packages/cctbx/eltbx/attenuation_coefficient.cpp`
  * `site-packages/cctbx/command_line/brehm_diederichs.py`
  * `site-packages/iotbx/xds/spot_xds.py`
  * `site-packages/iotbx/xds/correction.py`
  * `site-packages/iotbx/xds/blank_cbf.py`
  * `site-packages/iotbx/xds/xds_inp.py`
  * `site-packages/iotbx/xds/gain_cbf.py`
  * `site-packages/iotbx/xds/xds_cbf.py`
  * `site-packages/iotbx/xds/xparm.py`
  * `site-packages/iotbx/xds/geometry.py`
  * `site-packages/scitbx/glmtbx/glm.h`
  * `site-packages/scitbx/glmtbx/robust_glm.h`
  * `site-packages/scitbx/glmtbx/family.h`
  * `site-packages/scitbx/glmtbx/boost_python/ext.cc`

The different licenses should at least be aggregated in a large license file on the top level, showing which licenses apply where, to comply with all the required crediting. For the purposes of the conda recipe the package as a whole then needs to be declared as mixed-licensed.

The good news is though that fundamentally everything looks like it is legal to redistribute - apart from the Diamond files that don't contain any license reference, but I have it on good authority that those are fine 🙂 (and we should fix them upstream).

**Review by Anthchirp — COMMENTED** [review]
I had a look over the recipe and I have a few questions about declared dependencies and the licensing.

File: recipes/cctbx/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        - libsvm
+        - matplotlib-base
+        - mrcfile
+        - {{ pin_compatible('numpy') }}
+        - pillow
+        - psutil
+        - pyopengl
+        - python
+        - python.app  # [osx]
+        - reportlab
```
[the remark this answers is not in the document]
Comment (Anthchirp) [reply recipes/cctbx/meta.yaml:64]: I'll take that back, `rstbx` is packaged. Not sure why I missed that.

File: recipes/cctbx/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    - setuptools
+    - six
+
+outputs:
+  - name: cctbx-base
+    requirements:
+      run:
+        - biopython
+        - {{ pin_compatible('boost') }}
+        - boost-cpp
```
[the remark this answers is not in the document]
Comment (bkpoon) [reply recipes/cctbx/meta.yaml:50]: Yes, this package contains the headers, but it is also for some of the Boost libraries that we use. We discussed it earlier in the year (https://github.com/cctbx/dxtbx/pull/134#issuecomment-579464103). You're using [`boost-cpp`](https://github.com/dials/dials/blob/master/.conda-envs/linux.txt) for building DIALS, so you should be able to switch to this package once the recipe for DIALS is ready.

File: recipes/cctbx/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        - cython
```
[the remark this answers is not in the document]
Comment (bkpoon) [reply recipes/cctbx/meta.yaml:51]: Nice! This must have been left over from earlier tests with `cython`. I'll remove it once the build finishes.

**Comment by Anthchirp** [comment]
Looking through the files it looks like we are vendoring an older version of the ANTLR package in `site-packages/ucif/antlr3`. We apparently actually use this during the build process. I know that conda-forge does not like vendoring other packages but this one maybe could be an exception since ANTLR3 is not actually a conda-forge package (and has been out of support for years)

File: recipes/cctbx/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        - eigen
```
[the remark this answers is not in the document]
Comment (bkpoon) [reply recipes/cctbx/meta.yaml:52]: Yes, this can be removed since it should be a header only library.

File: recipes/cctbx/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        - future
+        - ipython
```
[the remark this answers is not in the document]
Comment (bkpoon) [reply recipes/cctbx/meta.yaml:54]: Users have requested that `ipython` be available by default, so it's one less step for general users.

**Comment by scopatz** [comment]
Yeah this is probably fine.  Do you need antlr at runtime?

File: recipes/cctbx/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        - libsvm
+        - matplotlib-base
+        - mrcfile
+        - {{ pin_compatible('numpy') }}
+        - pillow
+        - psutil
+        - pyopengl
```
[the remark this answers is not in the document]
Comment (bkpoon) [reply recipes/cctbx/meta.yaml:61]: Must have been something that was added and not pursued. Thanks!

File: recipes/cctbx/meta.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (bkpoon) [reply recipes/cctbx/meta.yaml:64]: If the `cctbx` package is going to be the base for some other packages, I basically built everything that can be used and deleted anything from `dxtbx` and `cbflib`.

File: recipes/cctbx/meta.yaml (hunk 5 above)
Comment (bkpoon) [reply recipes/cctbx/meta.yaml:85]: Some of these are a copyright, which is different from a license.

Also, something like `site-packages/scitbx/array_family/shared_plain.h` contain modifications allowed by the license and we follow the terms of the license in the file.
```
/* This code is derived in part from:
 *   STLport-4.5.3/stlport/stl/_vector.h
 *   STLport-4.5.3/stlport/stl/_vector.c
 *
 * Copyright (c) 1994
 * Hewlett-Packard Company
 *
 * Copyright (c) 1996,1997
 * Silicon Graphics Computer Systems, Inc.
 *
 * Copyright (c) 1997
 * Moscow Center for SPARC Technology
 *
 * Copyright (c) 1999
 * Boris Fomitchev
 *
 * This material is provided "as is", with absolutely no warranty expressed
 * or implied. Any use is at your own risk.
 *
 * Permission to use or copy this software for any purpose is hereby granted
 * without fee, provided the above notices are retained on all copies.
 * Permission to modify the code and to distribute modified code is granted,
 * provided the above notices are retained, and a notice that the code was
 * modified is included with the above copyright notice.
 *
 */
```
I'm fine with setting the license as mixed and adding the additional licenses.

**Comment by Anthchirp** [comment]
Oh, very good point - no. This is a build-time dependency only.

Coming to think about this we should probably split up the package (following the boost/boost-cpp model) and move all c/cpp/h source files into a cctbx-cpp output. That would also reduce the package size for a normal binary install.

**Comment by scopatz** [comment]
Yeah, it really doesn't matter if it is only a build time dep

**Comment by bkpoon** [comment]
@scopatz How do I include multiple license files?

@Anthchirp This isn't quite the same. `boost-cpp` includes headers and libraries. I don't think the library source files are included in the package. I can try removing the source files here. They are probably not be needed, but in some edge cases, I think a source file may be included when building a separate program. Let's get rid of them for now and see if any issues arise. We would not need `cctbx-cpp` package, though, since this would just be headers and libraries, just like `boost-cpp`.

**Comment by scopatz** [comment]
> How do I include multiple license files?

`licence_file` can be a list

**Comment by Anthchirp** [comment]
> @Anthchirp This isn't quite the same. `boost-cpp` includes headers and libraries. I don't think the library source files are included in the package. I can try removing the source files here. They are probably not needed, but in some edge cases, I think a source file may be included when building a separate program. Let's get rid of them for now and see if any issues arise. We would not need a `cctbx-cpp` package, though, since this would just be headers, libraries, and the Python files.

That's true, but the use case I'm thinking here is: If you only _run_ cctbx code all you need is the binary build artefacts and the python files. If you want to _compile against_ cctbx you'd need (at least) the correct header files (so maybe `cctbx-headers` rather than `-cpp`?). Having those in a package (or output) next to cctbx would be useful to compile a dxtbx conda package against, hopefully using standard conda-forge mechanisms, but keep the files out of the way if you don't explicitly need them. However, once that package is built these files are again not needed for an installation, it'll be the same for dials and cctbx+dxtbx headers.

I thought this is how `boost` <> `boost-cpp` related, hence my comment above, but it looks like one still has to contend with installing 13700 `.hpp` files even though they may not be used for anything 🙁. I'll ask that in a boost-cpp issue.

**Comment by bkpoon** [comment]
So generally, for non-Python packages, the `make install` step (or equivalent) will install headers and libraries. I guess linux distribution packages are done the way you described (e.g. `mesa-libGL` and `mesa-libGL-devel` or `python` and `python-devel`), but it does not look like the general policy for `conda` packages. Otherwise, we would also have a `python-devel` conda package in addition to just a `python` package. This sounds like a much broader discussion for the `conda-forge` community. For now, I think we stick with headers, libraries, and Python files.

Builds are still running, but I'll add lgpl-2.0, lgpl-2.1, lgpl-3, MIT, Boost, and wxWindows licenses in addition to our license once it finishes. The Diamond Light Source copyright is not a separate license, right?

**Comment by bkpoon** [comment]
An extra wrinkle is the `numpy` package (`conda` or `pip`). It has headers inside.

**Comment by Anthchirp** [comment]
Diamond files will be plain BSD-3-Clause without the Berkeley special bits.

**Comment by bkpoon** [comment]
@conda-forge-admin please rerender

**Comment by bkpoon** [comment]
I had to add back `mesa-libGLU-devel` in `yum_requirements.txt` because we include `glu.h` in some of the code. Is there a way similar to `libGL`? I had tried adding `- {{ cdt('mesa-libglu-devel') }}  # [linux]` in bcc258c75d93110b2411b2d1266d14d9ba7305ef, but that didn't work. Thanks!

**Comment by chrisburr** [comment]
Add `libglu` to ` host` and `{{ pin_compatible('libglu') }}` to `run`.

Also you're missing some yum_requirements for OpenGL: https://conda-forge.org/docs/maintainer/knowledge_base.html#libgl

**Comment by bkpoon** [comment]
@conda-forge-admin please rerender

**Comment by Anthchirp** [comment]
On installation this package sets up `phenix.*` commands. I thought these fall under a [different, non-free license](https://www.phenix-online.org/license/LICENSE) which is why we don't distribute those in eg. dials or ccp4?

**Comment by bkpoon** [comment]
Oh, those are just aliased names for some other commands (e.g. `phenix.xtriage` is exactly the same as `mmtbx.xtriage`). It's basically just branding so I added the flag to turn those off. The Phenix license only really applies to the code in the phenix repository which is not included here.

Any other suggestions or comments from anyone? The packages built from Azure Pipelines look good from my end (thanks @chrisburr for the `libglu` fix).

**Comment by Anthchirp** [comment]
I have a few more dispatcher questions.

The package generates loads of duplicate commands:
```
annlib.python
boost.python
ccp4io.python
cctbx.python
cctbx_website.python
cma_es.python
crys3d.python
fable.python
gltbx.python
iotbx.python
libtbx.python
mmtbx.python
omptbx.python
rstbx.python
scitbx.python
smtbx.python
spotfinder.python
wxtbx.python
```
these are basically aliases to `python`.

```
annlib.show_build_path
boost.show_build_path
ccp4io.show_build_path
cctbx.show_build_path
cctbx_website.show_build_path
cma_es.show_build_path
crys3d.show_build_path
fable.show_build_path
gltbx.show_build_path
iotbx.show_build_path
libtbx.show_build_path
mmtbx.show_build_path
omptbx.show_build_path
rstbx.show_build_path
scitbx.show_build_path
smtbx.show_build_path
spotfinder.show_build_path
wxtbx.show_build_path
          
annlib.show_dist_paths
boost.show_dist_paths
ccp4io.show_dist_paths
cctbx.show_dist_paths
cctbx_website.show_dist_paths
cma_es.show_dist_paths
crys3d.show_dist_paths
fable.show_dist_paths
gltbx.show_dist_paths
iotbx.show_dist_paths
libtbx.show_dist_paths
mmtbx.show_dist_paths
omptbx.show_dist_paths
rstbx.show_dist_paths
scitbx.show_dist_paths
smtbx.show_dist_paths
spotfinder.show_dist_paths
wxtbx.show_dist_paths
```
seem very redundant.

The package also sets up dispatchers with the name `pytest` and `py.test`, which will lead to conflicts with the `pytest` package. Those should be removed/not generated.

**Comment by Anthchirp** [comment]
Entry points aren't registered properly. For example:
```python
$ python -c "import pkg_resources;print(list(pkg_resources.iter_entry_points('pytest_randomly.random_seeder')))" 
[]
```
The expected output is:
```python
$ python -c "import pkg_resources;print(list(pkg_resources.iter_entry_points('pytest_randomly.random_seeder')))" 
[EntryPoint.parse('scitbx_flex = scitbx.array_family.flex:set_random_seed')]
```

**Comment by chrisburr** [comment]
https://conda.io/projects/conda-build/en/latest/resources/define-metadata.html#python-entry-points

**Comment by bkpoon** [comment]
The `python` aliases should be kept because users will have scripts or other packages that use these aliases.

The `show_dist_paths` can be changed later upstream, but I can manually delete them in this recipe.

The `py.test` and `pytest` aliases were added by you, and you had a pull request to remove them  (https://github.com/cctbx/cctbx_project/pull/488), but it broke something on your end. I can manually delete them in this recipe, but this should be fixed upstream.

The entry point was added by you (https://github.com/cctbx/cctbx_project/pull/383) and seems to be used for your testing infrastructure at Diamond Light Source. Is it necessary for general usage for most users? I explicitly skip the copying of the `egg-info` directories by default because of the documentation pointed out by @chrisburr.

**Comment by Anthchirp** [comment]
> The `python` aliases should be kept because users will have scripts or other packages that use these aliases.
> The `show_dist_paths` can be changed later upstream, but I can manually delete them in this recipe.
👍 

> The `py.test` and `pytest` aliases were added by you, and you had a pull request to remove them ([cctbx/cctbx_project#488](https://github.com/cctbx/cctbx_project/pull/488)), but it broke something on your end. I can manually delete them in this recipe, but this should be fixed upstream.
Yes, we can move that from cctbx to dxtbx for now. They are only required upstream because of the extra logic happening in the dispatchers to be able to find the modules directory. Regular pip-installed packages won't be able to find any cctbx modules in a development install. As I understand this won't be an issue for the conda package (as the sources are put in the right place), so we don't need to create it here.

As an aside, I am wondering - is the plan to eventually move to regular entry point declarations for all the commands that are being set up? Can we add some of the commands to the recipe test list?

> The entry point was added by you ([cctbx/cctbx_project#383](https://github.com/cctbx/cctbx_project/pull/383)) and seems to be used for your testing infrastructure at Diamond Light Source. Is it necessary for general usage for most users? I explicitly skip the copying of the `egg-info` directories by default because of the documentation pointed out by @chrisburr.

That is correct. The entry point is so that flex arrays can be correctly used in randomized tests of any downstream package. As this is a scitbx feature this should be set up in scitbx, not anywhere else. The documentation is a red herring - this is not a console_script entry point (which is what the documentation is dealing with), so it doesn't go into the recipe. To my knowledge there is no way to declare non-console_script entry points there. Maybe someone from conda-forge could suggest how this could be set?

Basically what we do in the development installation is roughly equivalent to
```python
setuptools.setup(
    name='libtbx.scitbx',
    description='libtbx entry point manager for scitbx',
    entry_points={
        "pytest_randomly.random_seeder": [
            "scitbx_flex = scitbx.array_family.flex:set_random_seed"
        ],
    },
)
```

**Comment by bkpoon** [comment]
> As an aside, I am wondering - is the plan to eventually move to regular entry point declarations for all the commands that are being set up?

That is not the plan since entry point declarations seem to hard-code an absolute path.

> Can we add some of the commands to the recipe test list?

Which commands are you interested in testing? These builds are already taking 2-3 hours per platform. As I mentioned in https://github.com/cctbx/cctbx_project/issues/451#issuecomment-654407571, I will be adding nightly builds of a `conda` package to our Azure Pipelines organization once this recipe is merged (modified to pull the latest sources). Then the nightly package can run more extensive tests since many of the tests require data in different repositories.

> To my knowledge there is no way to declare non-console_script entry points there. Maybe someone from conda-forge could suggest how this could be set?

There is https://conda.io/projects/conda-build/en/latest/resources/define-metadata.html#preserve-python-egg-directory. I did add the option to copy `.egg-info` directories for this reason (can be generalized to other egg-like directories). This can be fixed in a later release or build since this should not affect the vast majority of users.

**Comment by Anthchirp** [comment]
> > As an aside, I am wondering - is the plan to eventually move to regular entry point declarations for all the commands that are being set up?
> 
> That is not the plan since entry point declarations seem to hard-code an absolute path.

You mean the `#!` path? That's not really an issue - conda environments are not relocatable. I'm concerned about how custom dispatchers play with the toolchain to pack up conda environments for distribution. I ran a test with conda-pack and that went fine, however custom solutions always have the potential to require custom maintenance down the line.

> > Can we add some of the commands to the recipe test list?
> 
> Which commands are you interested in testing? These builds are already taking 2-3 hours per platform. As I mentioned in [cctbx/cctbx_project#451 (comment)](https://github.com/cctbx/cctbx_project/issues/451#issuecomment-654407571), I will be adding nightly builds of a `conda` package to our Azure Pipelines organization once this recipe is merged (modified to pull the latest sources). Then the nightly package can run more extensive tests since many of the tests require data in different repositories. The current import tests are to ensure that the Boost.Python wrapping is working.

I would suggest something simple like just `libtbx.show_commands`. This is a lightweight (<0.2 seconds) command that implicitly tests the env setup, something that is not covered by the import test. If anything were to go wrong with custom dispatchers in the future this could pick it up.

> > To my knowledge there is no way to declare non-console_script entry points there. Maybe someone from conda-forge could suggest how this could be set?
> 
> There is https://conda.io/projects/conda-build/en/latest/resources/define-metadata.html#preserve-python-egg-directory. I did add the option to copy `.egg-info` directories for this reason (can be generalized to other egg-like directories). This can be fixed in a later release or build since this should not affect the vast majority of users.

Sounds reasonable. I'm happy to take this on. Can you add me as a maintainer?

I believe I have tested the package extensively now, and I have nothing new to add 🙂

However, I have to go back to the `ipython` dependency discussion. I don't think "users expect this to be present" is a good enough argument to keep the dependency declared. Can users not equally run `conda install cctbx ipython`? Having it declared as a dependency _takes away_ the possibility of removing ipython (and its dependencies) from the environment, as that would break the environment consistency. Also - although I don't believe this to be the case for ipython or its dependencies - it may trigger unnecessary migrations in conda-forge which will cost CI build time and PR reviewer time.

**Comment by bkpoon** [comment]
> You mean the #! path? That's not really an issue - conda environments are not relocatable. I'm concerned about how custom dispatchers play with the toolchain to pack up conda environments for distribution. I ran a test with conda-pack and that went fine, however custom solutions always have the potential to require custom maintenance down the line.

We are not requiring `conda` for distribution or limiting builds to only work with `conda` as I mentioned here (https://github.com/cctbx/cctbx_project/issues/515#issuecomment-682367403). Any user can build cctbx as long as the dependencies are available in a `$PREFIX` directory.

> I would suggest something simple like just libtbx.show_commands.

That sounds good. An update has just been committed.

> Sounds reasonable. I'm happy to take this on. Can you add me as a maintainer?

Please bring it up for the next cctbx management meeting. I'll probably take a look next weekend to get this to work. It's much faster to iterate by running `conda build` locally on one version of Python (~ 15 minutes on my local linux server). Then test the full build across multiple platforms.

> However, I have to go back to the ipython dependency discussion. I don't think "users expect this to be present" is a good enough argument to keep the dependency declared.

Please raise an issue at https://github.com/cctbx/cctbx_project and/or the cctbxbb bulletin board (http://www.phenix-online.org/mailman/listinfo/cctbxbb) and/or the cctbx management meeting.

@chrisburr I think this can be merged if no one else on your end has any comments. Thanks everyone!

**Comment by Anthchirp** [comment]
> > However, I have to go back to the ipython dependency discussion. I don't think "users expect this to be present" is a good enough argument to keep the dependency declared.

As an immediate compromise - can this at least be moved to the `cctbx` output rather than `cctbx-base`?

**Comment by Anthchirp** [comment]
> We are not requiring `conda` for distribution or limiting builds to only work with `conda` as I mentioned here 

The entry points defined in `meta.yaml` only affect conda installations. How dispatchers are work elsewhere is beside the point.

> Please bring it up for the next cctbx management meeting.

Sure. Can you send me an invite please?

**Comment by Anthchirp** [comment]
I believe the most recent linter warning https://github.com/conda-forge/staged-recipes/pull/10021#issuecomment-695141637 is also still unresolved

**Comment by bkpoon** [comment]
I was following
https://github.com/conda-forge/hdf5-external-filter-plugins-feedstock/blob/9ec7a2236dec95ecd914ecf373d0643d5d576dcf/recipe/meta.yaml#L78-L81
for the `license` entry since we have multiple licenses.

Let me change it to an expression to see if it makes the linter happy.

**Comment by bkpoon** [comment]
@conda-forge-admin, please lint

**Comment by bkpoon** [comment]
@conda-forge-admin, please lint

**Comment by bkpoon** [comment]
@conda-forge-admin, please lint

**Comment by bkpoon** [comment]
@conda-forge-admin, please lint

**Comment by bkpoon** [comment]
@conda-forge-admin, please lint

**Comment by bkpoon** [comment]
@conda-forge-admin, please lint

**Comment by bkpoon** [comment]
@conda-forge-admin, please lint

**Comment by bkpoon** [comment]
@chrisburr The linter is happy with the `license` keyword now. It's nice that there is a specific entry for the LBNL modified version of BSD 3 clause.

@Anthchirp I have moved `ipython` to the `cctbx` output. And I have emailed the PIs about the next meeting.

**Comment by Anthchirp** [comment]
Thank you. I had a final look through the build logs, and these caught my eye:
```
2020-09-20T02:12:32.0981177Z WARNING (cctbx-base): run-exports library package conda-forge::libtiff-4.1.0-hc7e4089_6 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
2020-09-20T02:12:32.0983167Z WARNING (cctbx-base): run-exports library package conda-forge::hdf5-1.10.5-nompi_h3c11f04_1104 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```

Do we actually use `libtiff`? A quick search showed no results.

On hdf5: I believed that that was actually used, so I took another look. There is only a single reference in `rstbx/apps/stills/ersatz_nexus.cpp` and `.h`, which I thought was legitimate. But it turns out the warning is correct and that those files are not actually built or used - and they were [added by @graeme-winter back in 2012 for "safe keeping"](https://github.com/cctbx/cctbx_project/commit/08bff560105c17696e32a361198513f2e29d79e5), so they can probably be removed upstream.

**Comment by bkpoon** [comment]
@Anthchirp Actually, that helps with removing some other dependencies. I was building `dxtbx` (which is why `libtiff` is included) and `cbflib`, but it looks like I can avoid doing that. The `hdf5`, `libtiff`, and `msgpack-c` dependencies have been removed.

**Comment by bkpoon** [comment]
@chrisburr I think this is ready to be merged. Thanks!

File: recipes/cctbx/fix_macos_rpath.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+# =============================================================================
+if __name__ == '__main__':
+  exe_dev_files = glob.glob('build/exe_dev/*')
+  ext_files = glob.glob('build/lib/*_ext.so')
+  lib_files = glob.glob('build/lib/*.dylib')
+  test_files = glob.glob('build/**/tst_*', recursive=True) \
+               + glob.glob('build/**/timing/*', recursive=True) \
+               + glob.glob('build/**/boost_python/*.so', recursive=True)
+  for ext_file in exe_dev_files + ext_files + lib_files + test_files:
+    libraries = check_output(['otool', '-L', ext_file]).decode('utf8').split('\n')
```
Comment (chrisburr) [inline recipes/cctbx/fix_macos_rpath.py:20]: I think this should use `$OTOOL` and `$INSTALL_NAME_TOOL` as the conda-forge provided binaries are prefixed (last I checked it was `x86_64-apple-darwin13.4.0-install_name_tool`/`x86_64-apple-darwin13.4.0-otool` but you shouldn't rely on a fixed value).

```suggestion
    libraries = check_output([os.environ['OTOOL'], '-L', ext_file]).decode('utf8').split('\n')
```

File: recipes/cctbx/fix_macos_rpath.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+      lib = line.replace('\t', '').split()
+      if len(lib) > 0:
+        lib = lib[0]
+        new_lib = None
+        if lib.startswith('lib/'):
+          new_lib = os.path.join('@rpath', lib.split('/')[1])
+        elif lib.startswith('@rpath/lib/'):
+          new_lib = os.path.join('@rpath', lib.split('/')[2])
+        if new_lib is not None:
+          cmd = ['install_name_tool', '-change', lib, new_lib, ext_file]
```
Comment (chrisburr) [inline recipes/cctbx/fix_macos_rpath.py:31]:
```suggestion
          cmd = [os.environ["INSTALL_NAME_TOOL"], '-change', lib, new_lib, ext_file]
```

**Review by chrisburr — COMMENTED** [review]
Sorry just one last minor thing that's worth doing properly (especially considering the current macOS ARM transition).

**Comment by bkpoon** [comment]
Thanks! Good to know about the toolchain variables. And yes, we are interested in `osx-arm64` support. I should have access to some hardware once some final hardware is available (later this year?).
